extensions/blocks/story/edit.native.js

```diff
@@ -0,0 +1,301 @@
+		const { attributes, clientId } = this.props;
+
+		// TODO review states and dialog mechanism
+
+		// if ( this.state.isUploadInProgress ) {
+		// 	requestImageUploadCancelDialog( attributes.id );
+		// } else if (
+		// 	attributes.id &&
+		// 	getProtocol( attributes.src ) === 'file:'
+		// ) {
+		// 	requestImageFailedRetryDialog( attributes.id );
+		// }
+
+		// this.setState( {
+		// 	isCaptionSelected: false,
+		// } );
+
+		// TODO decide which course of action to take depending on current state for this Story block
+		// if ( isUploadInProgress ) {
+		// 	requestImageUploadCancelDialog( id );
+		// } else if ( shouldShowFailure ) {
+		// 	requestImageFailedRetryDialog( id );
+		// } else if ( isImage && url ) {
+		// 	requestImageFullscreenPreview( url );
+		// }
+		// requestMediaFilesEditorLoad( id );
+		// let's open the Story Creator and load this block in there
+		requestMediaFilesEditorLoad( attributes.mediaFiles, clientId );
+	}
+
+	// upload state handling methods
+	updateMediaUploadProgress( payload ) {
+		const { setAttributes } = this.props;
+		if ( payload.mediaUrl ) {
+			setAttributes( { url: payload.mediaUrl } );
+		}
+		if ( ! this.state.isUploadInProgress ) {
+			this.setState( { isUploadInProgress: true } );
+		}
+	}
+
+	finishMediaUploadWithSuccess( payload ) {
+		const { setAttributes } = this.props;
+		// setAttributes( { src: payload.mediaUrl, id: payload.mediaServerId } );
+		// find the mediaFiles item that needs to change via its id, and apply the new URL
+		// var updatedMediaFiles = this.replaceMediaUrlInMediaFilesById( payload.mediaId, payload.mediaUrl);
+		const updatedMediaFiles = this.replaceNewIdInMediaFilesByOldId(
+			payload.mediaId,
+			payload.mediaServerId,
+			payload.mediaUrl
+		);
+		setAttributes( { mediaFiles: updatedMediaFiles } );
+		this.setState( { isUploadInProgress: false } );
+	}
+
+	finishMediaUploadWithFailure( payload ) {
+		const { setAttributes } = this.props;
+		// TODO change this
+		// setAttributes( { id: payload.mediaId } );
+		this.setState( { isUploadInProgress: false } );
+	}
+
+	mediaUploadStateReset() {
+		const { setAttributes } = this.props;
+		// TODO change this
+		// setAttributes( { id: null, src: null } );
+		this.setState( { isUploadInProgress: false } );
+	}
+
+	// save state handling methods
+	updateMediaSaveProgress( payload ) {
+		const { setAttributes } = this.props;
+		if ( payload.mediaUrl ) {
+			setAttributes( { url: payload.mediaUrl } );
+		}
+		if ( ! this.state.isSaveInProgress ) {
+			this.setState( { isSaveInProgress: true } );
+		}
+	}
+
+	replaceMediaUrlInMediaFilesById( mediaId, mediaUrl ) {
+		const { attributes } = this.props;
+		if ( mediaId !== undefined && attributes.mediaFiles !== undefined ) {
+			for ( let i = 0; i < attributes.mediaFiles.length; i++ ) {
+				if ( attributes.mediaFiles[ i ].id === mediaId.toString() ) {
+					attributes.mediaFiles[ i ].url = mediaUrl;
+					attributes.mediaFiles[ i ].link = mediaUrl;
+				}
+			}
+		}
```

Comment:
nitpick, maybe a personal preference, but I was thinking this could work:
```
	attributes.mediaFiles.map(mediaFile => {
		if (mediaFile.id === mediaId.toString()) {
			return { ...mediaFile, url: mediaUrl, link: mediaUrl };
		}
		return mediaFile;
	});
```

Reply:
Thanks for bringing this up @guarani !  this is interesting :) 
As discussed elsewhere, I’ll take your suggestion and try applying it on the later PR here https://github.com/Automattic/jetpack/pull/17458/files#diff-0eb601171cf07f789577209ef0c3d0ede37c1c0aa66a59a851d457a7d4f3bb51R153  given I’m already taking care of making a copy of the mediaFiles array before trying to modify it, and then with the new copy I’m then setting it using setAttributes, which is needed given attributes should remain immutable and only changed via setAttributtes

Reply:
👋 @guarani  - applied your suggestion  in b885029e99 in PR #17458 👍